contracts/EtomicSwapV2.sol

```diff
@@ -0,0 +1,483 @@
+        TakerPaymentState state;
+    }
+
+    event TakerPaymentSent(bytes32 id);
+    event TakerPaymentApproved(bytes32 id);
+    event TakerPaymentSpent(bytes32 id, bytes32 secret);
+    event TakerPaymentRefundedSecret(bytes32 id, bytes32 secret);
+    event TakerPaymentRefundedTimelock(bytes32 id);
+
+    mapping(bytes32 => TakerPayment) public takerPayments;
+
+    address public immutable dexFeeAddress;
+
+    constructor(address feeAddress) {
+        require(feeAddress != address(0), "feeAddress must not be zero address");
+
+        dexFeeAddress = feeAddress;
+    }
+
+    function ethMakerPayment(
+        bytes32 id,
+        address taker,
+        bytes20 takerSecretHash,
+        bytes20 makerSecretHash,
+        uint32 paymentLockTime
+    ) external payable {
+        require(makerPayments[id].state == MakerPaymentState.Uninitialized, "Maker payment is already initialized");
+        require(taker != address(0), "Taker must not be zero address");
+        require(msg.value > 0, "ETH value must be greater than zero");
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                msg.value,
+                taker,
+                msg.sender,
+                takerSecretHash,
+                makerSecretHash,
+                address(0)
+            )
+        );
+
+        makerPayments[id] = MakerPayment(paymentHash, paymentLockTime, MakerPaymentState.PaymentSent);
+
+        emit MakerPaymentSent(id);
+    }
+
+    function erc20MakerPayment(
+        bytes32 id,
+        uint256 amount,
+        address tokenAddress,
+        address taker,
+        bytes20 takerSecretHash,
+        bytes20 makerSecretHash,
+        uint32 paymentLockTime
+    ) external {
+        require(makerPayments[id].state == MakerPaymentState.Uninitialized, "Maker payment is already initialized");
+        require(amount > 0, "Amount must not be zero");
+        require(taker != address(0), "Taker must not be zero address");
+
+        bytes20 paymentHash = ripemd160(
+            abi.encodePacked(
+                amount,
+                taker,
+                msg.sender,
+                takerSecretHash,
+                makerSecretHash,
+                tokenAddress
+            )
+        );
+
+        makerPayments[id] = MakerPayment(paymentHash, paymentLockTime, MakerPaymentState.PaymentSent);
+
+        emit MakerPaymentSent(id);
+
+        // Now performing the external interaction
+        IERC20 token = IERC20(tokenAddress);
+        // Ensure that the token transfer from the sender to the contract is successful
+        require(
+            token.transferFrom(msg.sender, address(this), amount),
+            "ERC20 transfer failed: Insufficient balance or allowance"
+        );
+    }
+
+    function spendMakerPayment(
+        bytes32 id,
+        uint256 amount,
+        address maker,
+        bytes20 takerSecretHash,
+        bytes32 makerSecret,
+        address tokenAddress
+    ) external {
```

Comment:
Q1: why in both `spend Taker/ Maker` methods we use `bytes32 makerSecret` param and `bytes20 takerSecretHash` (I mean only for taker sercret we provide hash format)?

Q2: and why both `Spend taker / maker payment` methods use `makerSecret`? and in both `refund Taker/Maker Payment Secrets` use `bytes32 takerSecret` ? I see, we dont change secret type(is it maker's or taker's) for Taker and Maker methods.

Reply:
Replied above